domain/removal/state/unit.go

```diff
@@ -0,0 +1,355 @@
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.UnitNotFound
+		} else if err != nil {
+			return errors.Errorf("running unit life query: %w", err)
+		}
+
+		return nil
+	})
+
+	return unitLife.Life, errors.Capture(err)
+}
+
+// DeleteUnit removes a unit from the database completely.
+func (st *State) DeleteUnit(ctx context.Context, unitUUID string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Get the net node UUID for the unit.
+	selectNetNodeStmt, err := st.Prepare(`
+SELECT    nn.uuid AS &entityUUID.uuid
+FROM      unit AS u
+LEFT JOIN net_node AS nn ON nn.uuid = u.net_node_uuid
+WHERE     u.uuid = $entityUUID.uuid;`, entityUUID{})
+	if err != nil {
+		return errors.Errorf("preparing unit net node query: %w", err)
+	}
+
+	unitUUIDRec := entityUUID{UUID: unitUUID}
+	deleteUnitStmt, err := st.Prepare(`
+DELETE FROM unit
+WHERE  uuid = $entityUUID.uuid;`, unitUUIDRec)
+	if err != nil {
+		return errors.Errorf("preparing unit delete: %w", err)
+	}
+
+	return errors.Capture(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var netNodeUUIDRec entityUUID
+		if err := tx.Query(ctx, selectNetNodeStmt, unitUUIDRec).Get(&netNodeUUIDRec); errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.UnitNotFound
+		} else if err != nil {
+			return errors.Errorf("getting net node UUID for unit %q: %w", unitUUID, err)
+		}
+
+		if err := st.deleteUnitAnnotations(ctx, tx, unitUUID); err != nil {
+			return errors.Errorf("deleting annotations for unit %q: %w", unitUUID, err)
+		}
+
+		if err := st.deleteCloudContainer(ctx, tx, unitUUID, netNodeUUIDRec.UUID); err != nil {
+			return errors.Errorf("deleting cloud container for unit %q: %w", unitUUID, err)
+		}
+
+		if err := st.deleteForeignKeyUnitReferences(ctx, tx, unitUUID); err != nil {
+			return errors.Errorf("deleting unit references for unit %q: %w", unitUUID, err)
+		}
+
+		if err := tx.Query(ctx, deleteUnitStmt, unitUUIDRec).Run(); err != nil {
+			return errors.Errorf("deleting unit for unit %q: %w", unitUUID, err)
+		}
+
+		return nil
+	}))
+}
+
+func (st *State) deleteUnitAnnotations(ctx context.Context, tx *sqlair.TX, uUUID string) error {
+	unitUUIDRec := unitUUID{UUID: uUUID}
+
+	deleteUnitAnnotationStmt, err := st.Prepare(`
+DELETE FROM annotation_unit
+WHERE  uuid = $unitUUID.unit_uuid`, unitUUIDRec)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	if err := tx.Query(ctx, deleteUnitAnnotationStmt, unitUUIDRec).Run(); err != nil {
+		return errors.Errorf("removing unit annotations: %w", err)
+	}
+	return nil
+}
+
+func (st *State) deleteCloudContainer(ctx context.Context, tx *sqlair.TX, uUUID, netNodeUUID string) error {
+	unitUUIDRec := unitUUID{UUID: uUUID}
+
+	if err := st.deleteCloudContainerPorts(ctx, tx, uUUID); err != nil {
+		return errors.Errorf("removing cloud container ports: %w", err)
+	}
+
+	if err := st.deleteCloudContainerAddresses(ctx, tx, netNodeUUID); err != nil {
```

Comment:
We only want to do this if the unit *has* a cloud container, which means we know it is not on a machine. Otherwise we will be deleting the machine addresses.

If the unit has a cloud container we want to delete the addresses, device and node.

Reply:
Righto.